### PR TITLE
StevenBlack-hosts: init at 2.5.52

### PR DESCRIPTION
**Pull request by pasqui23**
###### Motivation for this change
Adbloking based on @StevenBlack 's excellent hosts file

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/StevenBlack-hosts.nix
Comment (mmilata) [inline nixos/modules/services/networking/StevenBlack-hosts.nix]:
```suggestion
        Note that options thet modify directly the hosts file or the global DNS cache will error out as the hosts file is first built in the standard nix sandbox.
```

File: nixos/modules/services/networking/StevenBlack-hosts.nix
Comment (mmilata) [inline nixos/modules/services/networking/StevenBlack-hosts.nix]:
```suggestion
      description = "List of hosts to allow";
```

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (mmilata) [inline pkgs/applications/networking/StevenBlack-hosts/default.nix]: I'd suggest using the base32 hash format which is more prevalent in nixpkgs and used by tools such as `nix-prefetch-url`:
```
nix-hash --type sha256 --to-base32 "zQMzXci1/21PCvi8AGqyDataQl/kQJJ+jszxXd2XMQc="
```

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (mmilata) [inline pkgs/applications/networking/StevenBlack-hosts/default.nix]: `maintainer` field is missing. Probably a good time to add yourself to `maintainers-list.nix`:) Preferably in separate commit.

**Review by mmilata — COMMENTED** [review]
Looks quite good, I've added a few formal nitpicks to make review easier for someone who can actually merge. Also:
* I think it's preferred to have package and module updates in separate commits
* there's bunch of whitespace issues, perhaps try using one of the nix formatting tools?

File: nixos/modules/services/networking/StevenBlack-hosts.nix
Comment (Infinisil) [inline nixos/modules/services/networking/StevenBlack-hosts.nix]: This is a bit ugly here, I think this is IFD (import-from-derivation), where a derivation needs to be built during evaluation time. I just created https://github.com/NixOS/nixpkgs/pull/81945 to get around this, after which you will be able to do
```nix
{
  networking.hostFiles = [
    (pkgs.StevenBlack-hosts.override { /* ... */ })
  ];
}
```

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (Infinisil) [inline pkgs/applications/networking/StevenBlack-hosts/default.nix]: It might be a good idea to not put this in `$out` directly, but rather `$out/share/StevenBlack-hosts/hosts`. This then also allows people to install the package with `environment.systemPackages` or so, after which the hosts file will be in `/run/current-system/sw/share/StevenBlack-hosts/hosts`

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (Infinisil) [inline pkgs/applications/networking/StevenBlack-hosts/default.nix]: From the repo I see that the license is MIT

**Review by infinisil — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (pasqui23) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix]: Yeah,but the sources from which the repo get the files are not

https://github.com/StevenBlack/hosts#sources-of-hosts-data-unified-in-this-variant

File: nixos/modules/services/networking/StevenBlack-hosts.nix
Comment (Infinisil) [reply nixos/modules/services/networking/StevenBlack-hosts.nix]: Just merged above PR, so this can be done now

**Comment by pasqui23** [comment]
I have incorporated @Infinisil 's suggestions

File: nixos/modules/services/networking/StevenBlack-hosts.nix
Comment (Infinisil) [inline nixos/modules/services/networking/StevenBlack-hosts.nix]: What's the format of this option? This should be documented in the description. Also could there be a more appropriate type for this than just `lines`?

**Review by infinisil — CHANGES_REQUESTED** [review]
Also see @mmilata's review here: https://github.com/NixOS/nixpkgs/pull/80113#pullrequestreview-364169641

File: nixos/modules/services/networking/StevenBlack-hosts.nix
Comment (pasqui23) [reply nixos/modules/services/networking/StevenBlack-hosts.nix:28]: I've clarified the description and add an example so that it's clear how to use the option

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:6472]:
```suggestion
  stevenblack-hosts = callPackage ../applications/networking/StevenBlack-hosts { };
```

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/StevenBlack-hosts/default.nix:11]: Why is here a space?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (StevenBlack) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix]: The space character issue aside, this makes me think I should be publishing the hash for each hosts file I generate to ensure the integrity of the chain of custody.

@SuperSandro2000 who generates the hash currently?  It's most unusual to have a hash created downstream from its origin.￼  

Am I missing context here?

I am happy to generate the hashes in the future...￼

Comment (pasqui23) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix]: I have generated the hash, simply by running `nix-prefetch fetchFromGitHub …`.
I expect the update scripts to update said hash on version update.

It's pratically required for nixpkgs to include the hash of the source, as one of its objectives is to ensure reproducibility of the package builds.

Comment (StevenBlack) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix]: Thanks @pasqui23.

Another observation.  The latest release of hosts is [3.5.2](https://github.com/StevenBlack/hosts/releases/tag/3.5.2).  Release [2.5.52](https://github.com/StevenBlack/hosts/releases/tag/2.5.52), referenced here, is from Feb 9, 2020.  More than a year ago.  We release between 1 and 3 times per week, depending on what comes in.

Am I missing context here?

Comment (pasqui23) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix]: Adding an hash per host file would not be used by nixpkgs, as the hash already check the ccontents of *every* host file in the repo.

Comment (SuperSandro2000) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix]: > The space character issue aside, this makes me think I should be publishing the hash for each hosts file I generate to ensure the integrity of the chain of custody.

We do not verify such hashes normally and because of some normalization done fetchFromGitHub will most likely return different hashes anyway.

> who generates the hash currently? It's most unusual to have a hash created downstream from its origin.

That is totally normal for nix and hashes are verified by the CI and everyone who tries to build the packages locally because the hashes are not already in the binary cache.

> Another observation. The latest release of hosts is 3.5.2. Release 2.5.52, referenced here, is from Feb 9, 2020. More than a year ago. We release between 1 and 3 times per week, depending on what comes in.

The review process took some time so we will probably need to update it right after merging.

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/StevenBlack-hosts/default.nix:47]:
```suggestion
    cp . $odir/hosts
```

File: nixos/modules/services/networking/StevenBlack-hosts.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/StevenBlack-hosts.nix:40]:
```suggestion
        skipStaticHosts = mkDefault true;
```

Comment (pasqui23) [reply nixos/modules/services/networking/StevenBlack-hosts.nix:40]: No, as `flags` goes straight to the commandline flags of the build program (a bit like settings options in other modules).

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (pasqui23) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix:47]: This would copy everything from `.`, including many build files, in `$odir` and we only want `hosts`.

Comment (SuperSandro2000) [reply pkgs/applications/networking/StevenBlack-hosts/default.nix:47]:
```suggestion
    cp ./hosts $odir/hosts
```

like this?

**Comment by SuperSandro2000** [comment]
@pasqui23 can you fix the eval error?

@ofborg eval

File: pkgs/applications/networking/StevenBlack-hosts/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/StevenBlack-hosts/default.nix:59]:
```suggestion
    maintainers = [ mantainers.pasqui23 ];
```

**Pull request closed without merging**
